tests/integration/security/pass_through_filter_chain_test.go

```diff
@@ -52,206 +49,182 @@ func TestPassThroughFilterChain(t *testing.T) {
+					{
+						target: dSet[0],
+						port:   8085,
+						schema: protocol.HTTP,
+						want:   true,
+					},
+					{
+						target: dSet[0],
+						port:   8086,
+						schema: protocol.HTTP,
+						want:   true,
+					},
+					{
+						target: dSet[0],
+						port:   8087,
+						schema: protocol.TCP,
+						want:   true,
+					},
+					{
+						target: dSet[0],
+						port:   8088,
+						schema: protocol.TCP,
+						want:   true,
 					},
 				}
-				t.Run(name, func(t *testing.T) {
-					retry.UntilSuccessOrFail(t, func() error {
-						responses, err := from.ForwardEcho(context.TODO(), request)
-						if tc.want {
-							if err != nil {
-								return fmt.Errorf("want allow but got error: %v", err)
-							}
-							if len(responses) < 1 {
-								return fmt.Errorf("received no responses from request to %s", host)
-							}
-							if tc.schema == protocol.HTTP && response.StatusCodeOK != responses[0].Code {
-								return fmt.Errorf("want status %s but got %s", response.StatusCodeOK, responses[0].Code)
-							}
-						} else {
-							// Check HTTP forbidden response
-							if len(responses) >= 1 && response.StatusCodeForbidden == responses[0].Code {
-								return nil
-							}
-
-							if err == nil || !strings.Contains(err.Error(), "EOF") {
-								return fmt.Errorf("want error EOF but got: %v", err)
-							}
+				ctx.NewSubTest(fmt.Sprintf("In %s", cluster.Name())).Run(func(ctx framework.TestContext) {
+					for _, tc := range cases {
+						name := fmt.Sprintf("x->%s:%d[%t]", tc.target.Config().Service, tc.port, tc.want)
+						a := apps.A.Match(echo.InCluster(cluster)).Match(echo.Namespace(ns.Name()))[0]
+						from := getWorkload(a, t)
+						// The request should be handled by the pass through filter chain.
+						host := fmt.Sprintf("%s:%d", getWorkload(tc.target, t).Address(), tc.port)
+						request := &epb.ForwardEchoRequest{
+							Url:     fmt.Sprintf("%s://%s", tc.schema, host),
+							Count:   int32(callCount),
+							Message: "HelloWorld",
+							Headers: []*epb.Header{
+								{
+									Key:   "Host",
+									Value: host,
+								},
+							},
 						}
-						return nil
-					}, retry.Delay(250*time.Millisecond), retry.Timeout(30*time.Second))
+						ctx.NewSubTest(name).Run(func(ctx framework.TestContext) {
+							retry.UntilSuccessOrFail(t, func() error {
+								responses, err := from.ForwardEcho(context.TODO(), request)
+								if tc.want {
+									if err != nil {
+										return fmt.Errorf("want allow but got error: %v", err)
+									}
+									if len(responses) < 1 {
+										return fmt.Errorf("received no responses from request to %s", host)
+									}
+									if tc.schema == protocol.HTTP && response.StatusCodeOK != responses[0].Code {
+										return fmt.Errorf("want status %s but got %s", response.StatusCodeOK, responses[0].Code)
+									}
+									if ctx.Clusters().IsMulticluster() {
+										responses.CheckReachedClusters(ctx.Clusters())
```

Comment:
I don't fully understand why we get `xSet` at the top. Usually if you match a destination `echo.Instances` down to a smaller subset, you'd pass that here to validate you only hit certain clusters that the service is in. If you expect to hit all clusters, you don't need to match `apps.B` to only the instances in the same cluster – we'll be load balancing across all clusters.

Reply:
Yeah, looks like this getWorkload(tc.target, t).Address() always give the address of the first pod of the passed in target, which may not be resolved by echos in different network. But the response can be load balanced among all clusters.